source/Tests/MessageContractAnalyzerFixture.cs

```diff
@@ -0,0 +1,573 @@
+namespace Octopus.Core.Features.Something
+{
+    /// <summary>a command</summary>
+    public class {|#10:SomeCommand|}: ICommand<SomeCommand, SomeResponse> { }
+    /// <summary>a request</summary>
+    public class {|#11:SomeRequest|}: IRequest<SomeRequest, SomeResponse> { }
+    /// <summary>a response</summary>
+    public class {|#12:SomeResponse|}: IResponse { }
+
+    public class {|#13:SomeEvent|} : IEvent { }
+    public class {|#14:SomeResource|} : Resource { }
+}
+// should not flag on any of these
+namespace Octopus.Core.Features.ServerTasks.MessageContracts
+{
+    /// <summary>a command</summary>
+    public class SomeCommand: ICommand<SomeCommand, SomeResponse> { }
+    /// <summary>a request</summary>
+    public class SomeRequest: IRequest<SomeRequest, SomeResponse> { }
+    /// <summary>a response</summary>
+    public class SomeResponse : IResponse { }
+
+    public class SomeEvent : IEvent { }
+    public class SomeResource : Resource { }
+}
+// or these
+namespace Octopus.Core.MessageContracts
+{
+    /// <summary>a command</summary>
+    public class SomeCommand: ICommand<SomeCommand, SomeResponse> { }
+    /// <summary>a request</summary>
+    public class SomeRequest: IRequest<SomeRequest, SomeResponse> { }
+
+    /// <summary>a response</summary>
+    public class SomeResponse : IResponse { }
+    public class SomeEvent : IEvent { }
+    public class SomeResource : Resource { }
+}");
+
+            await Verify.VerifyAnalyzerAsync(source,
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(0),
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(1),
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(2),
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(3),
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(4),
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(10),
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(11),
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(12),
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(13),
+                new DiagnosticResult(Descriptors.ApiContractTypesMustLiveInTheAppropriateNamespace).WithLocation(14));
+        }
+
+        static string WithOctopusTypes(string source) => $"{Common.Usings}{source}{Common.MessageTypeDeclarations}";
+    }
+
+    static class Common
+    {
+        // Declarations copied from MessageContracts. We need the names and structure to match exactly but the implementations
```

Comment:
Is this so the tests work?

Reply:
Yeah. When you ask it to analyze some source with `IRequest` in it, it actually compiles that code, and the code will fail to compile if `IRequest` hasn't been defined anywhere.

The easiest way to solve this is to include some "mock" types; Existing analyzers written by other devs were using this pattern too, so I followed suit